### ISSUE TITLE
Clarify ElseExpression rule

### ISSUE THREAD

**Issue by sbesselsen** [phpmd/phpmd#288]
From the docs: "An if expression with an else branch is never necessary". It then goes on: "You can rewrite the conditions in a way that the else is not necessary and the code becomes simpler to read. To achieve this use early return statements."

I don't understand the reasoning. Why is an early return (or, I assume, continue) statement intrinsically simpler to read? Why should we avoid else statements simply because they are never necessary? (The if statement, I might add, is never necessary either!)

These factors are unchanged:
- There is still an if.
- The "else" condition is still evaluated in exactly the same cases, only now by "falling through" instead of branching to the else block. The rationale that the exact condition for evaluating the else code is not clear, persists.

New issues are introduced:
- When using early returns, postprocessing code that has to be run in both branches has to be duplicated, introducing the potential for later divergence.
- When using early returns, the "else" code is "promoted" to a shallower indentation level, suggesting to the casual reader that it is the default case, even if the cases may be 50/50.
- When moving the if/else into a separate method, that method then needs to use early returns as well, because it can't use else either. Does it make the code more readable to have that method simply for this single purpose?
- A ternary operator is often harder to read. The only advantage I see there is that it only allows expressions instead of statements.

I have had a spirited discussion with my coworkers, but fail to be convinced so far that this rule achieves its objectives. I understand that else statements often appear in bad code, but I see them as a symptom rather than a cause. I feel that encouraging less experienced developers to use early returns and ternary operators may cause them to write worse code, not better.

I have found only a few critiques of the else statement on Google, and all of them focus on bad uses of else, not on the use of else per se.

My suggestion is to either:
- Clarify some of these concerns in the docs, and improve the suggestions.
- Put this rule in the "controversial" set.

I haven't made my own version of the docs text yet, because first I would like to have a clear understanding of the reasoning behind the rule and make sure there is nothing I miss.

**Comment by ravage84** [comment]
I agree that the documentation could be improved to better point out the problems or better said potential problems.
And I personally agree it does not become more readable by default when not using ``else`` clauses.

> I feel that encouraging less experienced developers to use early returns and ternary operators may cause them to write worse code, not better.

As long as we point them to some ressources where they can become better developers, I see no problem.

In the end is phpmd for me personally not a tool to teach people how to be better programmers but to point them to some of the most obvious and common mistakes and pitfalls.
What we can do though is to direct them to ressources where they can read how to do it better...

@manuelpichler do you know of any good (PHP related at best) ressource which explains all the things about ``else`` clauses?

**Comment by TomBZombie** [comment]
One thing I have noticed with PHPMD and I'm not sure if this is intentional or not is that this will trigger the warning:

```php
if ($x) {}
else {
$statement1;
$statement2;
}

```

Whereas this will not:

```php
if ($x) {}
else $statement1;
```

I'm not sure if this is due to the brackets or the fact that this could be represented as a ternary statement so for consistency PHPMD does not flag it up

I think that in this case, where we're controlling program flow and calling different methods:

if ($x) $this->doX();
else $this->doY();

Should not flag up the error but if there was a bunch of code instead of just a function call it should flag it up. It looks like this is how PHPMD works at the moment.

I think the book Clean Code by Robert C. Martin describes this situation well:

> This implies that the blocks within if statements, else statements, while statements, and
so on should be one line long. Probably that line should be a function call. Not only does
this keep the enclosing function small, but it also adds documentary value because the
function called within the block can have a nicely descriptive name.
This also implies that functions should not be large enough to hold nested structures.
Therefore, the indent level of a function should not be greater than one or two. This, of
course, makes the functions easier to read and understand.

**Comment by ianchanning** [comment]
On p358 of Code Complete (2nd Edition) by Steve McConnell suggests that you *should* include else statements:

> **Consider the else clause**  If you think you need a plain *if* statement, consider whether you don't actually need an *if-then-else* statement. A classic General Motors analysis found that 50 to 80 percent of *if* statements should have had an *else* clause (Elshoff 1976)

> One option is to code the *else* clause - with a null statement if necessary - to show that the *else* clause has been considered. Coding null *elses* just to show that that case has been considered might be overkill, but at the very least, take the *else* case into account. When you have an *if* test without an *else*, unless the reason is obvious, use comments to explain why the *else* clause isn't necessary

I also can't see where in the [Original PMD rules](https://pmd.github.io/pmd-5.3.3/pmd-java/rules/index.html) there is any reference to completely re-writing functions to exclude else statements.

I second @sbesselsen 's request to put this in the controversial list.

**Comment by Jurigag** [comment]
I think ElseExpression should only happen if there is return or throw inside of it, then there are really no point of having this return/throw in else. Otherwise it should be allowed.

**Comment by ravage84** [comment]
Agree with @Jurigag.

Improving the docs *and* only raise a violation if the else is really superfluous, seems to me the right way.

This way, the rule can stay in clean code and should not be considered controversial.

**Comment by Jurigag** [comment]
Also there is some case like this:

```php
if (some condition) {
$a = 1;
$b = 2;
}
else {
$a = null;
$b = null;
}
```

Some people would say that else is necessary. Well we can remove it by just setting $a and $b to null before if, but not sure if it's good way, for me code like this would look kind of weird, not sure how to handle something like this.

**Comment by ravage84** [comment]
Since I guess the detection could become quite complex for more advanced situation, I currently would only raise a violation for clear situations like with a return or throw.

We then could add another rule like ``AdvancedElseExpression`` or simply ``ElseExpression`` but in the controversional folder...

**Comment by Jurigag** [comment]
Yea, but still there should be check if we can simplify if to ternary operator.

**Comment by ravage84** [comment]
I don't se the benefit of this check. Or better said, I don't see the relaton between tenary operator and the current ElseExpression rule. Can you elaborate more on that?

**Comment by Jurigag** [comment]
I mean currently if we have if else which can be simplified to ternary operator is obviously detected. But if we want to change it to some limited area like throw/return i think it should check too if we can simplify to ternary operator. I mean conditional operator/shorthand if as well(`?` as well `?:`)

**Comment by ravage84** [comment]
I'm going to close this as I will try to improve the docs as a first step.

For interested parties, here is [an article](http://williamdurand.fr/2013/06/03/object-calisthenics/#2-dont-use-the-else-keyword) why else clauses are bad and what alternatives you have, though it certainly does not cover everything.

**Comment by fuzzy76** [comment]
The ternary operator _is_ an else-clause, it just has a different syntax. Recommending its use is admitting else is ok.

**Comment by ianchanning** [comment]
@ravage84 sorry old post but that article only has the following reasoning for not using else clauses:

> Do you remember the last time you saw a nested conditional? Did you enjoy reading it? I don’t think so, and that is exactly why it should be avoided.

That's a fairly weak statement for getting rid of a fundamental concept in programming languages. I personally agree and I do prefer [guard clauses](https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html), but there should be better arguments put forward.

**Comment by 626Pilot** [comment]
TL;DR: This rule is under "clean code", but the Clean Code book uses else statements in MANY locations. **This rule does not belong here. It belongs in Controversial.**

"An if expression with an else branch is never necessary."

Neither is PHP. Let's write all our web apps in assembly language.

What's that? You want to use PHP because it works better for you than assembly language? It's more expressive and easy to understand and maintain?

I don't write else statements for the compiler. I write them for myself, and for all the programmers who will be reading and maintaining my code after me. They express intention, which is among the kindest things I can do for everyone who has reason to understand what I'm "else"ing.

It is possible to use early returns. They are not "better" than else statements. At the end of the day, the behavior is identical either way.

If you prefer guard statements, or Rube-Goldberg-like code contraptions that allow you to avoid using 'else', please, be my guest. That preference does not obligate the rest of us to stop using one of the most obvious statements of intent that a programmer can possibly use, or to tolerate endless squawking about it, or to pollute our PHPDoc blocks with overrides that make it harder to read the source in exchange for exactly nothing.

"Else" is not prohibited in PSR-2. Instead, they show a three-way if statement with two elses right at the top of the document:
```
<?php
namespace Vendor\Package;

use FooInterface;
use BarClass as Bar;
use OtherVendor\OtherPackage\BazClass;

class Foo extends Bar implements FooInterface
{
    public function sampleMethod($a, $b = null)
    {
        if ($a === $b) {
            bar();
        } elseif ($a > $b) {
            $foo->bar($arg1);
        } else {
            BazClass::bar($arg2, $arg3);
        }
    }

    final public static function bar()
    {
        // method body
    }
}
```
How is this not just someone's arbitrary decision, rolled into phpmd's behavior with no actual research or discussion? It's not in PSR-2, and I couldn't find it anywhere in Clean Code. **In fact, Clean Code uses else statements in MANY locations.**

**Comment by Jurigag** [comment]
I'm still thinking that my first comment here 2 years ago is a way to go, it could stay in clean code only if it would consider return/throw statement in else condition.

**Comment by sbesselsen** [comment]
Interesting to see that this discussion is still going. I have learned a lot in 3 years, as I'm sure we all have, but I still see no reason to disallow else statements. All arguments against it that anyone here has been able to put forward boil down to "you can use it in a bad way", which is true of literally every other language feature.

Personally I would say this: leaving this check in is a strongly opinionated stance. If phpmd as a project is willing to be strongly opinionated, simply state it in the docs. "We discourage else because we feel that it can lead to bad code." Just don't state it as some kind of common knowledge or folk wisdom. I can live with it, as I have been doing for the past 3 years ;)

I understand that this is low priority, but surely the maintainers would like to remove this issue from the queue?

**Comment by bagermen** [comment]
`
if ($we_wrote_weird_thing) {
    $close_discussion();
} else {
    $remove_it();
}
`

**Comment by ThomasLabstep** [comment]
> This rule does not belong here. It belongs in Controversial.

I agree.

Look at this situation:
```
$user = $this->em->getRepository(User::class)->findOneBy([
    'email' => $email,
]);
if ($user) {
    $this->io->text("Found existing User with email: $email");
} else {
    // Create user
}

return $user;
```

PHPMD wants:
```
$user = $this->em->getRepository(User::class)->findOneBy([
    'email' => $email,
]);
if ($user) {
    $this->io->text("Found existing User with email: $email");

   return $user;
}
// Create User

return $user;
```

Is it more readable ? Maybe.
Do I need to add an extra return statement. Yes.

**Comment by kylekatarnls** [comment]
No PHPMD do not "want" you to do this, you have other alternatives:

```php
$user = $this->em->getRepository(User::class)->findOneBy([
    'email' => $email,
]);

$this->ensureUserExists($user, $email);

return $user;

...
public function ensureUserExists($user)
{
  if ($user) {
    $this->io->text("Found existing User with email: $email");

   return;
  }

  // Create User
}
```

And so in this case, you have a more readable code and 2 methods where each one has a more restrictive responsibility (you can explicit with the method name).

When you have an else, you should not ask yourself: "How to convert this code 1-to-1 to early-return?" You should ask yourself: "Can it be split into more atomic steps?"

And you will discover this rules help you to create methods with better defined responsibilities.

**Comment by ThomasLabstep** [comment]
> "Can it be split into more atomic steps?"

Understood.

@kylekatarnls Thanks!

**Comment by asbjornu** [comment]
I agree with @Jurigag. If each clause in an `if/elseif/else` statement ends with a `return` or `throw`, it makes sense to just turn them all into separate `if` statements. But when there is no `return` from the statements, but they instead do assignment, logging, etc., it makes no sense to discourage the use of `else`.

**Comment by kylekatarnls** [comment]
> it makes no sense to discourage the use of else

It does: here is the reason (and example that do no imply throw/return):
https://medium.com/@kylekatarnls/stop-using-else-84c1fa977202

I write a lot of code every day, and I didn't use `else` for years and I can ensure you my code is cleaner this way.

Code rules of PHPMD or any static analyzers are common practices in some language/company/region communities, there is no absolute universal rules that would benefit 100% of coding people, it's up to you to select the rules you need.
